dash/development/validator.py

```diff
@@ -0,0 +1,112 @@
+from textwrap import dedent
+import plotly
+import cerberus
+
+
+class DashValidator(cerberus.Validator):
+    types_mapping = cerberus.Validator.types_mapping.copy()
+    types_mapping.pop('list')  # To be replaced by our custom method
+    types_mapping.pop('number')  # To be replaced by our custom method
+
+    def _validator_plotly_figure(self, field, value):
+        if not isinstance(value, (dict, plotly.graph_objs.Figure)):
+            self._error(
+                field,
+                "Invalid Plotly Figure: Not a dict")
+        if isinstance(value, dict):
+            try:
+                plotly.graph_objs.Figure(value)
+            except (ValueError, plotly.exceptions.PlotlyDictKeyError) as e:
+                self._error(
+                    field,
+                    "Invalid Plotly Figure:\n\n{}".format(e))
+
+    def _validator_options_with_unique_values(self, field, value):
+        if not isinstance(value, list):
+            self._error(field, "Invalid options: Not a dict!")
+        values = set()
+        for i, option_dict in enumerate(value):
+            if not isinstance(option_dict, dict):
+                self._error(
+                    field,
+                    "The option at index {} is not a dictionary!"
+                    .format(i)
+                )
+            if 'value' not in option_dict:
+                self._error(
+                    field,
+                    "The option at index {} does not have a 'value' key!"
+                    .format(i)
+                )
+            curr = option_dict['value']
+            if curr in values:
+                self._error(
+                    field,
+                    ("The options list you provided was not valid. "
+                     "More than one of the options has the value {}."
+                     .format(curr))
+                )
+            values.add(curr)
+
+    def _validate_type_list(self, value):
+        if isinstance(value, list):
+            return True
+        elif isinstance(value, (self.component_class, str)):
+            return False
+        try:
+            value_list = list(value)
+            if not isinstance(value_list, list):
+                return False
+        except (ValueError, TypeError):
+            return False
+        return True
+
+    # pylint: disable=no-self-use
+    def _validate_type_number(self, value):
+        if isinstance(value, (int, float)):
+            return True
+        if isinstance(value, str):  # Since int('3') works
+            return False
```

Comment:
The lines after will cast to int/float if it's a string, but that check before that.

Reply:
This logic is actually to handle numpy ints and floats without having to add `numpy` as a run-time dependency just for type checking.
```
>>> isinstance(np.int(7), int)
True
>>> isinstance(np.float(7), float)
True
>>> isinstance(np.uint8(7), int)
False
>>> isinstance(np.int64(7), int)
True
>>> isinstance(np.int32(7), int)
False
>>> isinstance(np.float64(7), float)
True
>>> isinstance(np.float32(7), float)
False
```
But all numpy int types are castable to an `int`, and same for floats.

Reply:
This would mean that `'7'` or `'0.7` would not be validated as `propType.number` even though they would have been automatically cast to an integer in the past. Not sure if we should also accept these? I think with `propTypes.number`, you should be using numbers in Python just to stay consistent, but this is more of a stylistic opinion rather than a requirement for correct functionality.

Reply:
We should not accept a string for a number, even if it's castable. Maybe add a comment that it is for the numpy types, because reading that is not so obvious and it look like a logic error.